### PR TITLE
[apollo-api] Parse Longs only once

### PR DESCRIPTION
**Pull request by martinbonnin**
~First (baby) step toward removing `BigDecimal` as a mandatory intermediate when converting from json to GraphQL `Int` and `Double`: Remove the default adapters for Float and Long.~

Edit: After some investigation, this pull request keeps the `DEFAULT_ADAPTERS` for now as that would be a breaking change. It still optimizes the parsing of `Longs`. There is no need to parse from String again since `BufferedSourceJsonReader` already parses internally. 

Follow up issue about removing `DEFAULT_ADAPTERS` there: https://github.com/apollographql/apollo-android/issues/2486

### CODE REVIEW

**Comment by sav007** [comment]
I have a question regarding `Long`, looking into the code of our json reader seems like we read long values `https://github.com/apollographql/apollo-android/blob/master/apollo-api/src/commonMain/kotlin/com/apollographql/apollo/api/internal/json/BufferedSourceJsonReader.kt#L545`

**Comment by martinbonnin** [comment]
> I have a question regarding Long, looking into the code of our json reader seems like we can read long values

It can indeed but it's unused in the default GraphQL type system. It could be useful in custom scalar type adapters but they should be registered manually and not put in `DEFAULT_ADAPTERS`? 
More generally, what do we need `DEFAULT_ADAPTERS` for? All the builtin GraphQL types are handled in a different path (`readString`, `readInt`, etc...). The only case I could find for `DEFAULT_ADAPTERS` is to handle the `ID`
 scalar type but that's one in many default adapters.

**Comment by martinbonnin** [comment]
I added a couple of benchmarks. It turns out that going from `Int` -> `Double` -> `Int` is slower than `Int` -> `BigDecimal` -> `Int`. So ideally, we should avoid using intermediate representations, which make sense.

**Comment by martinbonnin** [comment]
And I just pushed an optimization that divides by ~2 the time spent on `Int` parsing. The code that we had from Moshi already parses the `Longs` as it scans the Json and we were doing it a second time just after.

**Comment by sav007** [comment]
>More generally, what do we need DEFAULT_ADAPTERS for?

They provide more convenience for user, when for instance some custom scalar type mapped to `java.lang.Long` there is no need to provide adapter for it. Same for `Float`.

Is there any reason to remove them?

File: apollo-integration/src/test/java/com/apollographql/apollo/performance/NumberParsingTest.kt [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                        0.until(10000).map {
+                          JsonPrimitive(random.nextInt())
+                        }
+                    )
+                )
+            )
+        )
+    )
+
+    val json = data.toString().encodeUtf8()
```
Comment (sav007) [inline apollo-integration/src/test/java/com/apollographql/apollo/performance/NumberParsingTest.kt:55]: Is there any reason to not use our `JsonWriter` to compose thus test json? I mean at least we can avoid extra dependency to `kotlinx-serialization` that is used as far as I understood only for this test?

**Comment by martinbonnin** [comment]
> They provide more convenience for user, when for instance some custom scalar type mapped to java.lang.Long there is no need to provide adapter for it. Same for Float.

Got it 👍 . Yea, removing these adapters will be a breaking change for users that rely on them.

> Is there any reason to remove them?

There's an argument to be made for API consistency. Some adapters like `java.util.Date` currently have to be manually added using `addCustomTypeAdapter` while others like `com.apollographql.apollo.api.FileUpload` are registered automatically. 

We could instead declare the implementations and let the user register them as needed:

```kotlin
// build.gradle.kts
apollo {
   customTypeMapping.set(mapOf("TimeStamp" to "java.lang.Long"))
}

// Apollo.kt
apolloClient = ApolloClient.builder().addCustomTypeAdapter(
                                   CustomType.TimeStamp, 
                                   ApolloCustomTypeAdapter.LongAdapter
                             )
```

We could also add `Iso8601DateAdapter`, I think this is one of the first custom scalars that users encounter.

File: apollo-integration/src/test/java/com/apollographql/apollo/performance/NumberParsingTest.kt (hunk 1 above)
Comment (martinbonnin) [reply apollo-integration/src/test/java/com/apollographql/apollo/performance/NumberParsingTest.kt]: It was mostly me being lazy to count the number of `beginObject` and `endObjects` using the `JsonWriter` API. I just did it.

**Comment by martinbonnin** [comment]
@sav007 I restored the default adapters for now.  Follow up issue there: https://github.com/apollographql/apollo-android/issues/2486

**Pull request merged**